arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,248 @@
+        runtime->dispatch(pid, sender, payload.get());
+      } else {
+        fmt::print("Error serializing UnknownMessage");
+        std::abort();
+      }
+    }
+  }
+
+  auto finish() -> void override { finished.store(true); }
+  auto finishedAndIdle() -> bool override {
+    return finished.load() and idle.load();
+  }
+
+  auto serialize() -> velocypack::SharedSlice override {
+    auto res = inspection::serializeWithErrorT(*this);
+    if (not res.ok()) {
+      std::abort();
+    }
+    return res.get();
+  }
+
+ private:
+  void push(ActorPID sender, typename Config::Message&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+  void push(ActorPID sender, ActorError&& msg) {
+    pushToQueueAndKick(std::make_unique<InternalMessage>(
+        sender,
+        std::make_unique<MessageOrError<typename Config::Message>>(msg)));
+  }
+
+  void kick() {
+    // Make sure that *someone* works here
+    (*runtime->scheduler)([self = this->weak_from_this()]() {
+      auto me = self.lock();
+      if (me != nullptr) {
+        me->work();
+      }
+    });
+  }
+
+  void work() {
+    if (finished.load() == true) {
+      return;
+    }
+    auto i = batchSize;
+
+    while (auto msg = inbox.pop()) {
+      state = std::visit(
+          typename Config::template Handler<Runtime>{
+              {pid, msg->sender, std::move(state), runtime}},
+          msg->payload->item);
+      i--;
+      if (i == 0) {
+        // push more work to scheduler if queue is still not empty
+        if (not inbox.empty()) {
+          kick();
+          return;
+        }
+        break;
+      }
+    }
+
+    idle.store(true);
+
+    // push more work to scheduler if a message was added to queue after
+    // previous inbox.empty check
+    auto isIdle = true;
+    if (not inbox.empty() and idle.compare_exchange_strong(isIdle, false)) {
+      kick();
+    }
+  }
+
+  struct InternalMessage
+      : arangodb::pregel::mpscqueue::MPSCQueue<InternalMessage>::Node {
+    InternalMessage(
+        ActorPID sender,
+        std::unique_ptr<MessageOrError<typename Config::Message>>&& payload)
+        : sender(sender), payload(std::move(payload)) {}
+    ActorPID sender;
+    std::unique_ptr<MessageOrError<typename Config::Message>> payload;
+  };
+  template<typename Inspector>
+  auto inspect(Inspector& f, InternalMessage& x) {
+    return f.object(x).fiels(f.field("sender", x.sender),
+                             f.field("payload", x.payload));
+  }
+
+  auto pushToQueueAndKick(std::unique_ptr<InternalMessage> msg) -> void {
+    inbox.push(std::move(msg));
+
+    // only push work to scheduler if actor is idle (meaning no work is waiting
+    // on the scheduler and no work is currently processed in work())
+    auto isIdle = idle.load();
+    if (isIdle and idle.compare_exchange_strong(isIdle, false)) {
+      kick();
+    }
+  }
+
+ public:
```

Comment:
why? It looks like details implementation
```suggestion
 private:
```

```suggestion
 protected:
```